python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py

```diff
@@ -382,9 +381,20 @@ def evaluate_for_asset(
     ) -> "AssetConditionResult":
         from .asset_condition.asset_condition import AssetConditionResult
 
-        new_asset_partitions_to_request = self.get_new_asset_partitions_to_request(context)
+        missed_ticks = self.missed_cron_ticks(context)
+        new_asset_partitions = self.get_new_candidate_asset_partitions(context, missed_ticks)
+
+        # if it's the first time evaluating this rule, must query for the actual subset that has
+        # been materialized since the previous cron tick, as materializations may have happened
+        # before the previous evaluation, which
+        # `context.materialized_requested_or_discarded_since_previous_tick_subset` would not capture
+        if context.previous_evaluation is None:
```

Comment:
why is this check here and

`not context.previous_evaluation or not context.previous_evaluation_timestamp`

above?

Can we just add a `is_first_tick` property and be consistent?

Reply:
This is all basically in service of appeasing the type system -- either one of those checks would be sufficient, it's just a matter of telling the type system that neither of those values will be None in the other branch.

I could definitely add an `is_first_tick` property, my concern is just that we'll then need to do `check.not_none` all over the place in order to "prove" that those values will not be None if is_first_tick is `False`, which seems at least a bit more risky than this setup.

Reply:
> This is all basically in service of appeasing the type system

I won't push back on this anymore. But my recommendation is if you are appeasing the typesystem only, then we should do a separate assert to do, to make it clear to code reader that that is what is going on.

In general case if you have

```python
if a.biz_logic() and a.some_prop_only_truthy_if_biz_logic_is_true:
    a.some_prop_only_truthy_if_biz_logic_is_true.do_something()
```

The following is preferable    

```python
if a.biz_logic():
    assert a.some_prop_only_truthy_if_biz_logic_is_true # for type checker
    a.some_prop_only_true_if_biz_logic_is_true.do_something()
```

In the future if someone screwed and changed `some_prop_only_truthy_if_biz_logic_is_true` or `biz_logic` to violate previous invariant, the bug is way more obvious.

Being very explicit about the assumptions one is making when coding (sometimes at the expense of terseness) is very useful